activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb

```diff
@@ -1815,7 +1815,23 @@ def foreign_key_name(table_name, options)
 
         def foreign_key_for(from_table, **options)
           return unless use_foreign_keys?
-          foreign_keys(from_table).detect { |fk| fk.defined_for?(**options) }
+
+          if options[:_skip_column_match]
+            return foreign_keys(from_table).detect { |fk| fk.defined_for?(**options) }
+          end
+
+          keys = foreign_keys(from_table)
+
+          if options[:column]
+            expected_columns = Array(options[:column]).map(&:to_s)
+            keys.select! { |fk| Array(fk.options[:column]).map(&:to_s) == expected_columns }
+          elsif options[:to_table] && !options.key?(:primary_key) && keys.size > 1
+            default_column = foreign_key_column_for(options[:to_table], "id").to_s
+            match = keys.find { |fk| Array(fk.options[:column]).map(&:to_s) == [default_column] }
+            keys = [match].compact if match
+          end
+
+          keys.detect { |fk| fk.defined_for?(**options) }
```

Comment:
Haven't tested, but something like this is shorter and should work too
```suggestion
          keys = foreign_keys(from_table)
          
          if keys.size == 1 || options[:_skip_column_match]
            keys.find { |fk| fk.defined_for?(**options) }
          elsif options[:column]
            expected_columns = Array(options[:column]).map(&:to_s)
            keys = keys.select { |fk| Array(fk.column) == expected_columns }
            keys.find { |fk| fk.defined_for?(**options) }
          else
            default_column = foreign_key_column_for(options[:to_table], "id")
            keys.find { |fk| fk.column == default_column }
          end
```

Reply:
That works, thanks for tidying it up!
Fixed in 02bcd2e